### DISCUSSION TITLE
engine.execute() always tries to format string even without params in MySQL+pymysql

### DISCUSSION THREAD

**Discussion by JWSong** [sqlalchemy/sqlalchemy#5857] [Usage Questions] [answered]
# Describe the bug

engine.execute() always tries to format string even without params in MySQL+pymysql. This causes an error when executing statements including '%' character. 

# To Reproduce

```py
import pymysql

engine = create_engine(db_uri)
result = engine.execute('SELECT "%"').fetchone()
print(result) 
```

## Expected
```py
('%',)
```

## Error

```py
Connected to pydev debugger (build 201.6668.115)
Traceback (most recent call last):
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1514, in _handle_dbapi_exception
    util.raise_(exc_info[1], with_traceback=exc_info[2])
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/pymysql/cursors.py", line 161, in execute
    query = self.mogrify(query, args)
  File "/Users/jsong/PycharmProjects/collector/venv/lib/python3.8/site-packages/pymysql/cursors.py", line 140, in mogrify
    query = query % self._escape_args(args, conn)
ValueError: unsupported format character '"' (0x22) at index 9
```

## Versions.
 - OS: macOS Catalina 10.15.6
 - Python: 3.9
 - SQLAlchemy: 1.3.20
 - Database: MySQL
 - DBAPI: pymysql

# Additional comment

I have tested only two databases, MySQL and SQLite. I got this error in MySQL only, specifically with pymysql DBAPI. I found out that sqlalchemy.engine.default.DefaultExecutionContext._init_statement sets `parameters` to [{}] when args are not provided. And pymysql library checks args like below

```py
if args is not None:
    # tries to format the statement by using '%' operator
```
instead of

```py
if args:
   ....
```

I think this issue is more on pymysql but I have decided report this issue in both repositories. How do you think about this issue?

**Thank you!**

**Comment by gordthompson** [comment, marked as answer]
Hi @JWSong .

First off, the `engine.execute()` method is considered legacy as of version 1.4 and will be removed in version 2.0.

Also, passing a raw string to `.execute()` is deprecated. We should either

1. wrap the string using the `text()` function, or
2. use the `exec_driver_sql()` method of a `Connection` object.

Using option (1) works "out of the box":

``` python
import sqlalchemy as sa

# …

sql = 'SELECT "%"'
result = engine.execute(sa.text(sql)).fetchone()
print(result)  # ('%',)
```

or, better:

``` python
import sqlalchemy as sa

# …

sql = 'SELECT "%"'
with engine.begin() as conn:
    result = conn.execute(sa.text(sql)).fetchone()
print(result)  # ('%',)
```

For option (2) we need to tweak the SQL command so the DBAPI won't mis-interpret the `%` character:

``` python
sql = 'SELECT "%%"'
with engine.begin() as conn:
    result = conn.exec_driver_sql(sql).fetchone()
print(result)  # ('%',)
```

**Comment by zzzeek** [comment]
SQLAlhcemy makes a decision here to pass parameters, including an empty collection, to cursor.execute() in all cases so that the sender of the SQL string (usually SQLAlchemy itself) does not have to guess if there will be parameters or not and hence whether or not the string form of the statement has to change.   when using exec_driver_sql() with a format (i.e. "%s" ) driver simply double up the percent signs.

**Comment by gordthompson** [comment]
> 
> 
> when using exec_driver_sql() with a format (i.e. "%s" ) driver simply double up the percent signs.

Right. This is documented here

https://mysqlclient.readthedocs.io/user_guide.html#mysqldb

where it says

"Note that any literal percent signs in the query string passed to execute() must be escaped, i.e. %%."

**Comment by gordthompson** [comment]
To clarify, with a raw pymysql connection,

``` python
crsr.execute('SELECT "%"')
```

works, while

``` python
crsr.execute('SELECT "%"', [])
```

throws "TypeError: not enough arguments for format string", whereas

``` python
crsr.execute('SELECT "%%"', [])
```

works as in the first example.

**Comment by JWSong** [comment]
Thank you all for your responses!
